lib/codecs/src/decoding/framing/chunked_gelf.rs

```diff
@@ -0,0 +1,715 @@
+        // | 64 bits    | 8 bits          | 8 bits       | remaining bits       |
+        // +------------+-----------------+--------------+----------------------+
+        //
+        // As this codec is oriented for UDP, the chunks (datagrams) are not guaranteed to be received in order,
+        // nor to be received at all. So, we have to store the chunks in a buffer (state field) until we receive
+        // all the chunks of a message. When we receive all the chunks of a message, we can concatenate them
+        // and return the complete payload.
+
+        // We need 10 bytes to read the message id, sequence number and total chunks
+        // if chunk.remaining() < 10 {
+        //     let src_display = format!("{chunk:?}");
+        //     warn!(message = "Received malformed chunk headers (message ID, sequence number and total chunks) with less than 10 bytes. Ignoring it.",
+        //         src = src_display,
+        //         remaining = chunk.remaining(),
+        //         internal_log_rate_limit = true
+        //     );
+        //     return Ok(None);
+        // }
+
+        ensure!(
+            chunk.remaining() >= 10,
+            InvalidChunkHeaderSnafu { header: chunk }
+        );
+
+        let message_id = chunk.get_u64();
+        let sequence_number = chunk.get_u8();
+        let total_chunks = chunk.get_u8();
+
+        ensure!(
+            total_chunks > 0 && total_chunks <= GELF_MAX_TOTAL_CHUNKS,
+            InvalidTotalChunksSnafu {
+                message_id,
+                sequence_number,
+                total_chunks
+            }
+        );
+
+        ensure!(
+            sequence_number < total_chunks,
+            InvalidSequenceNumberSnafu {
+                message_id,
+                sequence_number,
+                total_chunks
+            }
+        );
+
+        let mut state_lock = self.state.lock().expect("poisoned lock");
+
+        ensure!(
+            state_lock.len() < self.pending_messages_limit,
+            PendingMessagesLimitReachedSnafu {
+                message_id,
+                sequence_number,
+                pending_messages_limit: self.pending_messages_limit
+            }
+        );
+
+        let message_state = state_lock.entry(message_id).or_insert_with(|| {
+            // We need to spawn a task that will clear the message state after a certain time
+            // otherwise we will have a memory leak due to messages that never complete
+            let state = Arc::clone(&self.state);
+            let timeout = self.timeout;
+            let timeout_handle = tokio::spawn(async move {
+                tokio::time::sleep(timeout).await;
+                let mut state_lock = state.lock().unwrap();
+                if state_lock.remove(&message_id).is_some() {
+                    let message = format!("Message was not fully received within the timeout window of {}ms. Discarding it.",
```

Comment:
As we are not returning any error from here, there are not recorded metrics about dropped that in this case... is this the correct way to handle this data discard?